File: sdk/typescript/src/regulatory_dependency.ts
/**
 * OSMP Regulatory Dependency Grammar — Rule 8 for validateComposition
 * Validates instruction chains against REQUIRES rules loaded from MDR corpora.
 * Dependency rules are SAL expressions using the same glyph operators as
 * the instructions they govern.
 *
 * Patent ref: OSMP-001-UTIL Claim 40 (pending)
 * License: Apache 2.0
 */

import { readFileSync, existsSync } from "fs";
import type { CompositionIssue } from "./validate.js";

// ─── Data Structures ────────────────────────────────────────────────────────

export interface DependencyRule {
  entry: string;        // e.g. "F:BVLOS[P]"
  namespace: string;    // e.g. "F"
  opcode: string;       // e.g. "BVLOS"
  slotValue: string;    // e.g. "P" or "" if no slot
  requiresRaw: string;  // e.g. "REQUIRES:F:REMID[S]∨F:REMID[M]"
  alternatives: string[][];  // parsed: [[prereq_pattern, ...], ...]
}

// ─── Parser ─────────────────────────────────────────────────────────────────

const PREREQ_RE = /([A-Z]{1,2}):([A-Z][A-Z0-9]*)(?:\[([^\]]+)\])?/;

export function parseRequiresExpression(requires: string): string[][] {
  let expr = requires;
  if (expr.startsWith("REQUIRES:")) expr = expr.slice(9);
  // Split on ∨ (OR), then split each alternative on ∧ (AND) for conjunctive prerequisites.
  // Result: [[conjunct, ...], ...] — at least ONE group where ALL conjuncts satisfied.
  return expr.split("\u2228")  // ∨
    .map(p => p.trim())
    .filter(p => p.length > 0)
    .map(p => p.split("\u2227").map(c => c.trim()).filter(c => c.length > 0));
}

// ─── MDR Loader ─────────────────────────────────────────────────────────────

export function loadMDRDependencyRules(mdrPath: string): DependencyRule[] {
  if (!existsSync(mdrPath)) return [];

  const lines = readFileSync(mdrPath, "utf-8").split("\n");
  const rules: DependencyRule[] = [];
  let inSectionB = false;

  for (const line of lines) {
    const stripped = line.trim();
    if (stripped.includes("SECTION B")) { inSectionB = true; continue; }
    if (stripped.startsWith("SECTION ") && !stripped.includes("SECTION B")) {
      if (inSectionB) break;
    }
    if (!inSectionB) continue;
    if (!stripped || stripped.startsWith("Format:") ||
        stripped.startsWith("===") || stripped.startsWith("---") ||
        stripped.startsWith("Note:") || stripped.startsWith("Dependency rules")) {
      continue;
    }

    const parts = stripped.split(",");
    if (parts.length < 5 || !parts[0].includes(":")) continue;

    const depRule = (parts[4] || "").trim();
    if (!depRule.startsWith("REQUIRES:")) continue;

    const nsOp = parts[0].trim();
    const slotValue = parts[1].trim();
    const nsParts = nsOp.split(":");
    if (nsParts.length < 2) continue;

    const ns = nsParts[0];
    const opcode = nsParts[1];
    const entry = slotValue ? `${ns}:${opcode}[${slotValue}]` : `${ns}:${opcode}`;

    rules.push({
      entry,
      namespace: ns,
      opcode,
      slotValue,
      requiresRaw: depRule,
      alternatives: parseRequiresExpression(depRule),
    });
  }
  return rules;
}

// ─── Chain Frame Extraction ─────────────────────────────────────────────────

const CHAIN_FRAME_RE = /([A-Z]{1,2}):([A-Z][A-Z0-9]*)(?:\[([^\]]+)\]|:([A-Z0-9][A-Z0-9_.]+))?/g;

export function extractChainFrames(sal: string): { frames: Set<string>; opcodes: Set<string> } {
  const frames = new Set<string>();
  const opcodes = new Set<string>();
  const re = new RegExp(CHAIN_FRAME_RE.source, "g");
  let m: RegExpExecArray | null;

  while ((m = re.exec(sal)) !== null) {
    const [, ns, opcode, bracketVal, colonVal] = m;
    opcodes.add(`${ns}:${opcode}`);
    const val = bracketVal || colonVal;
    if (val) frames.add(`${ns}:${opcode}[${val}]`);
  }
  return { frames, opcodes };
}

// ─── Prerequisite Matching ──────────────────────────────────────────────────

function prereqSatisfied(pattern: string, frames: Set<string>, opcodes: Set<string>): boolean {
  const m = PREREQ_RE.exec(pattern);
  if (!m) return false;
  const [, ns, opcode, slot] = m;
  if (slot) return frames.has(`${ns}:${opcode}[${slot}]`);
  return opcodes.has(`${ns}:${opcode}`);
}

function checkAlternatives(alts: string[][], frames: Set<string>, opcodes: Set<string>): boolean {
  for (const group of alts) {
    if (group.every(prereq => prereqSatisfied(prereq, frames, opcodes))) {
      return true;
    }
  }
  return false;
}

// ─── Rule 8: Regulatory Dependency Validation ───────────────────────────────

export function validateRegulatoryDependencies(
  sal: string,
  rules: DependencyRule[],
): CompositionIssue[] {
  if (rules.length === 0) return [];

  const { frames, opcodes } = extractChainFrames(sal);

  // Build lookup
  const lookup = new Map<string, DependencyRule>();
  for (const rule of rules) {
    lookup.set(rule.entry, rule);
    if (!rule.slotValue) {
      lookup.set(`${rule.namespace}:${rule.opcode}`, rule);
    }
  }

  const issues: CompositionIssue[] = [];

  for (const frame of frames) {
    const rule = lookup.get(frame);
    if (rule && !checkAlternatives(rule.alternatives, frames, opcodes)) {
      const reqDisplay = rule.requiresRaw.replace("REQUIRES:", "");
      issues.push({
        rule: "REGULATORY_DEPENDENCY",
        severity: "error",
        message: `${rule.entry} requires ${reqDisplay} as a regulatory prerequisite. The prerequisite is absent from the instruction chain.`,
        frame: rule.entry,
      });
    }
  }

  for (const bare of opcodes) {
    if (frames.has(bare)) continue;
    const rule = lookup.get(bare);
    if (rule && !rule.slotValue && !checkAlternatives(rule.alternatives, frames, opcodes)) {
      const reqDisplay = rule.requiresRaw.replace("REQUIRES:", "");
      issues.push({
        rule: "REGULATORY_DEPENDENCY",
        severity: "error",
        message: `${rule.entry} requires ${reqDisplay} as a regulatory prerequisite. The prerequisite is absent from the instruction chain.`,
        frame: rule.entry,
      });
    }
  }

  return issues;
}

File: sdk/go/osmp/regulatory_dependency.go
package osmp

// Regulatory Dependency Grammar — Rule 8 for ValidateComposition
// Validates instruction chains against REQUIRES rules loaded from MDR corpora.
// Dependency rules are SAL expressions using the same glyph operators as
// the instructions they govern.
//
// Patent ref: OSMP-001-UTIL Claim 40 (pending)
// License: Apache 2.0

import (
	"fmt"
	"os"
	"strings"
)

// DependencyRule represents a single regulatory prerequisite from the MDR.
type DependencyRule struct {
	Entry       string     // e.g. "F:BVLOS[P]"
	Namespace   string     // e.g. "F"
	Opcode      string     // e.g. "BVLOS"
	SlotValue   string     // e.g. "P" or "" if no slot
	RequiresRaw string     // e.g. "REQUIRES:F:REMID[S]∨F:REMID[M]"
	Alternatives [][]string // parsed: [[prereq_pattern, ...], ...]
}

// SAL regex building blocks live in sal_patterns.go.
// Local alias preserves the existing name without changing call sites.
var prereqRe = salPrereqRe

// ParseRequiresExpression parses "REQUIRES:F:REMID[S]∧F:AV[Part107]∨F:REMID[M]" into alternatives.
// Each ∨-split alternative is further split on ∧ for conjunctive prerequisites.
// Result: [[conjunct, ...], ...] — at least ONE group where ALL conjuncts satisfied.
func ParseRequiresExpression(requires string) [][]string {
	expr := requires
	if strings.HasPrefix(expr, "REQUIRES:") {
		expr = expr[9:]
	}
	parts := strings.Split(expr, "\u2228") // ∨
	var result [][]string
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}
		conjuncts := strings.Split(p, "\u2227") // ∧
		var group []string
		for _, c := range conjuncts {
			c = strings.TrimSpace(c)
			if c != "" {
				group = append(group, c)
			}
		}
		if len(group) > 0 {
			result = append(result, group)
		}
	}
	return result
}

// LoadMDRDependencyRules loads regulatory dependency rules from an MDR CSV.
func LoadMDRDependencyRules(mdrPath string) ([]DependencyRule, error) {
	data, err := os.ReadFile(mdrPath)
	if err != nil {
		return nil, err
	}

	var rules []DependencyRule
	lines := strings.Split(string(data), "\n")
	inSectionB := false

	for _, line := range lines {
		stripped := strings.TrimSpace(line)
		if strings.Contains(stripped, "SECTION B") {
			inSectionB = true
			continue
		}
		if strings.HasPrefix(stripped, "SECTION ") && !strings.Contains(stripped, "SECTION B") {
			if inSectionB {
				break
			}
		}
		if !inSectionB {
			continue
		}
		if stripped == "" || strings.HasPrefix(stripped, "Format:") ||
			strings.HasPrefix(stripped, "===") || strings.HasPrefix(stripped, "---") ||
			strings.HasPrefix(stripped, "Note:") || strings.HasPrefix(stripped, "Dependency rules") {
			continue
		}

		parts := strings.SplitN(stripped, ",", 8)
		if len(parts) < 5 || !strings.Contains(parts[0], ":") {
			continue
		}

		depRule := strings.TrimSpace(parts[4])
		if !strings.HasPrefix(depRule, "REQUIRES:") {
			continue
		}

		nsOp := strings.TrimSpace(parts[0])
		slotValue := strings.TrimSpace(parts[1])
		nsParts := strings.SplitN(nsOp, ":", 2)
		if len(nsParts) < 2 {
			continue
		}

		ns := nsParts[0]
		opcode := nsParts[1]
		entry := ns + ":" + opcode
		if slotValue != "" {
			entry = fmt.Sprintf("%s:%s[%s]", ns, opcode, slotValue)
		}

		rules = append(rules, DependencyRule{
			Entry:        entry,
			Namespace:    ns,
			Opcode:       opcode,
			SlotValue:    slotValue,
			RequiresRaw:  depRule,
			Alternatives: ParseRequiresExpression(depRule),
		})
	}
	return rules, nil
}

// chain frame extraction regex: captures bracket [VAL] and colon :VAL notation
// chainFrameRe aliases the shared salChainFrameRe from sal_patterns.go.
var chainFrameRe = salChainFrameRe

// ExtractChainFrames extracts all frames from a SAL instruction chain.
// Returns (frames with slots normalized to bracket notation, bare opcodes).
func ExtractChainFrames(sal string) (map[string]bool, map[string]bool) {
	frames := map[string]bool{}
	opcodes := map[string]bool{}

	for _, m := range chainFrameRe.FindAllStringSubmatch(sal, -1) {
		ns, opcode := m[1], m[2]
		bracketVal, colonVal := m[3], m[4]
		opcodes[ns+":"+opcode] = true
		val := bracketVal
		if val == "" {
			val = colonVal
		}
		if val != "" {
			frames[fmt.Sprintf("%s:%s[%s]", ns, opcode, val)] = true
		}
	}
	return frames, opcodes
}

func prereqSatisfied(pattern string, frames, opcodes map[string]bool) bool {
	m := prereqRe.FindStringSubmatch(pattern)
	if m == nil {
		return false
	}
	ns, opcode, slot := m[1], m[2], m[3]
	if slot != "" {
		return frames[fmt.Sprintf("%s:%s[%s]", ns, opcode, slot)]
	}
	return opcodes[ns+":"+opcode]
}

func checkAlternatives(alts [][]string, frames, opcodes map[string]bool) bool {
	for _, group := range alts {
		allSatisfied := true
		for _, prereq := range group {
			if !prereqSatisfied(prereq, frames, opcodes) {
				allSatisfied = false
				break
			}
		}
		if allSatisfied {
			return true
		}
	}
	return false
}

// ValidateRegulatoryDependencies checks an instruction chain against MDR
// dependency rules. Returns CompositionIssue list (empty if all satisfied).
func ValidateRegulatoryDependencies(sal string, rules []DependencyRule) []CompositionIssue {
	if len(rules) == 0 {
		return nil
	}

	frames, opcodes := ExtractChainFrames(sal)

	// Build lookup
	lookup := map[string]*DependencyRule{}
	for i := range rules {
		lookup[rules[i].Entry] = &rules[i]
		if rules[i].SlotValue == "" {
			lookup[rules[i].Namespace+":"+rules[i].Opcode] = &rules[i]
		}
	}

	var issues []CompositionIssue

	for frame := range frames {
		if rule, ok := lookup[frame]; ok {
			if !checkAlternatives(rule.Alternatives, frames, opcodes) {
				reqDisplay := strings.TrimPrefix(rule.RequiresRaw, "REQUIRES:")
				issues = append(issues, CompositionIssue{
					Rule:     "REGULATORY_DEPENDENCY",
					Severity: "error",
					Message: fmt.Sprintf(
						"%s requires %s as a regulatory prerequisite. "+
							"The prerequisite is absent from the instruction chain.",
						rule.Entry, reqDisplay),
					Frame: rule.Entry,
				})
			}
		}
	}

	for bare := range opcodes {
		if _, inFrames := frames[bare]; inFrames {
			continue
		}
		if rule, ok := lookup[bare]; ok {
			if rule.SlotValue != "" {
				continue
			}
			if !checkAlternatives(rule.Alternatives, frames, opcodes) {
				reqDisplay := strings.TrimPrefix(rule.RequiresRaw, "REQUIRES:")
				issues = append(issues, CompositionIssue{
					Rule:     "REGULATORY_DEPENDENCY",
					Severity: "error",
					Message: fmt.Sprintf(
						"%s requires %s as a regulatory prerequisite. "+
							"The prerequisite is absent from the instruction chain.",
						rule.Entry, reqDisplay),
					Frame: rule.Entry,
				})
			}
		}
	}

	return issues
}
